Translation - Don't call proptypes (#85)
reactjs/fr.reactjs.org: content/warnings/dont-call-proptypes.md

---

> title: Avertissement Don't Call PropTypes

Reviewer: ```suggestion
title: Avertissement : don't call PropTypes
```

On va préfixer les titres par « Avertissement » en mode catégoriel, donc deux-points. Ensuite, on doit rester sur la VO pour faciliter l'identification avec le message pondu par la lib, mais ne pas succomber au Title Case de la VO pour autant.  Ici toutefois `PropTypes` est en fait un élément de code, donc on conserve sa casse.

layout: single
permalink: warnings/dont-call-proptypes.html
---

> > Note :

Reviewer: ```suggestion
N’appelez pas les PropTypes

> Remarque
```

- La première ligne non-remarque de la page sert de sous-titre, c'est un bon endroit pour fournir une traduction du message d'avertissement.
- cf. Glossaire. On retire en général les deux-points par ailleurs.

>

> > `React.PropTypes` a été déplacé dans un paquet différent depuis React v15.5. Merci d'utiliser [la bibliothèque `prop-types` à la place](https://www.npmjs.com/package/prop-types).

Reviewer: ```suggestion
> `React.PropTypes` a été déplacé dans un paquet dédié depuis React v15.5. Merci d'utiliser désormais [la bibliothèque `prop-types`](https://www.npmjs.com/package/prop-types).
```

- Ici “different” indique plus « distinct du paquet principal ».
- On évite « à la place » comme la peste (cf. Glossaire sur “instead”)

>
>Nous fournissons [un script codemod](/blog/2017/04/07/react-v15.5.0.html#migrating-from-react.proptypes) pour automatiser la conversion.

> Dans une future version majeure de React, le code implémentant les fonctions de validation PropType sera supprimé en production. Une fois que cela se produit, tout code appelant ces fonctions manuellement (qui n'est pas supprimé en production) générera une erreur.

Reviewer: ```suggestion
Dans une future version majeure de React, le code implémentant les fonctions de validation de PropType sera supprimé en production. Quand ça arrivera, tout code appelant ces fonctions manuellement (si ce code n'est pas supprimé en production) lèvera une erreur.
```

> ### Déclarer des PropTypes est toujours correct {#declaring-proptypes-is-still-fine}

Reviewer: ```suggestion
### Déclarer des PropTypes reste une bonne idée {#declaring-proptypes-is-still-fine}
```

Attention à « toujours » qui peut vouloir dire “still” (« encore ») ou “always”, ce qui n'a pas le même sens. Une reformulation s'impose souvent. Ensuite, ce n'est pas juste « correct » mais franchement conseillé 😉

> L'utilisation normale de PropTypes est toujours prise en charge :

Reviewer: ```suggestion
L'utilisation normale de PropTypes reste prise en charge :
```

```javascript
Button.propTypes = {
  highlighted: PropTypes.bool
};
```

Rien ne change ici.

### N'appelez pas PropTypes directement {#dont-call-proptypes-directly}

> L'utilisation de PropTypes autrement que par l'annotation de composants React n'est plus soutenu :

Reviewer: ```suggestion
L'utilisation de PropTypes autrement que par l'annotation de composants React n'est plus acceptée :
```

« Soutenu » ?!  Ce n'est pas “support” comme dans “support group” ou “looking for support”, mais comme “support a feature / use case”, donc « prise en charge » (le franglais « supporter » étant un barbarisme, puisque le véritable sens français de « supporter » est soit « servir de support physique » soit « accepter / tolérer » quelque chose de désagréable).

Ici toutefois il s'agit moins d'une prise en charge (comme « prise en charge native par les navigateurs ») que d'une acceptation, car on parle d'une mauvaise pratique dès le départ, recourir à « accepté(e) » (par React) est plus fort dans ce contexte.

```javascript
var apiShape = PropTypes.shape({
  body: PropTypes.object,
  statusCode: PropTypes.number.isRequired
}).isRequired;

> // N'est plus soutenu !

Reviewer: ```suggestion
// N'est plus accepté !
```

var error = apiShape(json, 'response');
```

> Si vous comptez utiliser de tels PropTypes, nous vous encourageons à utiliser ou créer un fork de PropTypes (tels que [ces](https://github.com/aackerman/PropTypes) [deux](https://github.com/developit/proptypes) paquets).

Reviewer: ```suggestion
Si vous avez besoin d’utiliser PropTypes de cette façon, nous vous encourageons à utiliser ou créer un *fork* de PropTypes (tels que [ces](https://github.com/aackerman/PropTypes) [deux](https://github.com/developit/proptypes) modules).
```

On insiste sur le côté « besoin » et pas juste « envie »

Si vous ne corrigez pas l'avertissement, ce code plantera en production avec React 16.

### Si vous n'appelez pas directement PropTypes mais recevez toujours l'avertissement {#if-you-dont-call-proptypes-directly-but-still-get-the-warning}

> Inspectez la trace de pile générée par l'avertissement. Vous trouverez la définition de composant responsable de l'appel direct PropTypes. Le problème est probablement dû à des PropTypes tiers qui enveloppent les PropTypes de React, par exemple :

Reviewer: ```suggestion
Inspectez la pile d’appels générée par l'avertissement. Vous y trouverez la définition de composant responsable de l'appel direct à PropTypes. Le problème est probablement dû à des PropTypes tiers qui enrobent ceux de React, par exemple :
```

}
```

> Dans ce cas, `ThirdPartyPropTypes.deprecated` est un container appelant `PropTypes.bool`. Ce modèle en soi est correct, mais déclenche un faux positif, car React pense que vous appelez directement PropTypes. La section suivante explique comment résoudre ce problème pour une bibliothèque implémentant quelque chose similaire à `ThirdPartyPropTypes`. Si ce n'est pas une bibliothèque que vous avez écrite, vous pouvez créer un problème à ce sujet.

Reviewer: ```suggestion
Dans ce cas, `ThirdPartyPropTypes.deprecated` est un enrobage qui appelle `PropTypes.bool`. Ce modèle n’a rien de mauvais en soi, mais il déclenche un faux positif car React pense que vous appelez directement PropTypes. La section suivante explique comment résoudre ce problème pour une bibliothèque implémentant quelque chose de similaire à `ThirdPartyPropTypes`. Si ce n'est pas une bibliothèque que vous avez écrite, vous pouvez déposer une *issue* sur son dépôt public.
```

- Wrapper = Enrobage
- Attention au recours excessif à « correct », parfaitement… correct 😉 en français québécois mais connoté très différemment dans un français plus large.
- Ici on parle clairement d'une *issue* au sens GitHub, attention donc à une traduction trop littérale.

Author: Mon dieu c'est terrible, l'influence Québécoise... je ne m'en même plus compte !

> ### Correction du faux positif dans les PropTypes tiers {#fixing-the-false-positive-in-third-party-proptypes}

Reviewer: ```suggestion
### Corriger le faux positif des PropTypes tiers {#fixing-the-false-positive-in-third-party-proptypes}
```

Toujours préférer un infinitif, plus entraînant, pour les gérondifs en début de titre

> Si vous êtes l'auteur d'une bibliothèque PropTypes tierce et que vous laissez les consommateurs encapsuler les PropTypes React existants, ils peuvent commencer à voir cet avertissement provenant de votre bibliothèque. Cela est dû au fait que React ne voit pas le dernier argument "secret" qu'il [passe](https://github.com/facebook/react/pull/7132) pour détecter les appels manuels PropTypes.

Reviewer: ```suggestion
Si vous êtes l'auteur·e d'une bibliothèque PropTypes tierce et que vous laissez vos utilisateurs enrober les PropTypes React existants, ils risquent de voir apparaître cet avertissement provenant de votre bibliothèque. C’est dû au fait que React ne voit pas le dernier argument « secret » qu'il [passe](https://github.com/facebook/react/pull/7132) pour détecter les appels manuels à PropTypes.
```

> Voici comment y remédier. Nous utiliserons `deprecated` à partir de  [react-bootstrap/react-prop-types](https://github.com/react-bootstrap/react-prop-types/blob/0d1cd3a49a93e513325e3258b28a82ce7d38e690/src/deprecated.js) comme exemple. L'implémentation actuelle ne transmet que les arguments `props`, `propName`, et `componentName` :

Reviewer: ```suggestion
Voici comment y remédier. Nous utiliserons le `deprecated` de  [react-bootstrap/react-prop-types](https://github.com/react-bootstrap/react-prop-types/blob/0d1cd3a49a93e513325e3258b28a82ce7d38e690/src/deprecated.js) comme exemple. L'implémentation actuelle ne transmet que les arguments `props`, `propName`, et `componentName` :
```

Très bien, le « y remédier » 👍

}
```

> Afin de corriger le faux positif, assurez-vous de transmettre **tous** les arguments au PropType encapsulé. Chose facile à faire avec la notation ES6 `... rest`:

Reviewer: ```suggestion
Afin de corriger le faux positif, assurez-vous de transmettre **tous** les arguments au PropType enrobé. C’est facile à faire avec la notation ES6 `...rest` :
```

```javascript
export default function deprecated(propType, explanation) {

>   return function validate(props, propName, componentName, ...rest) { // Notez ...rest ici

Reviewer: ```suggestion
  return function validate(props, propName, componentName, ...rest) { // Remarquez le ...rest ici
```

      }
    }

    return propType(props, propName, componentName, ...rest); // et ici
  };
}
```

> Cela fera taire l'avertissement.

Reviewer: ```suggestion
Ça fera taire l'avertissement.
```

Le « fera taire » est nickel 💯 mais on évite « Cela », trop formel (cf. glossaire)
